File: windows/DeusKVM.Companion.Core/ClipboardTransfer.cs
using System.Buffers.Binary;
using System.Text;

namespace DeusKVM.Companion.Core;

// One requested block at a time keeps clipboard traffic below the BLE queue cap.
// Epochs isolate desktop/link lifetimes; sequences isolate individual copies.
public sealed class ClipboardTransfer(Action<Protocol.Message, byte[]> send, Action<byte[]> apply)
{
    public const int MaximumBytes = 65536, BlockBytes = 1024;
    private static readonly UTF8Encoding Utf8 = new(false, true);
    private uint epoch, sequence;
    public uint Sequence => sequence;
    private byte[]? local;
    private uint? seenOffer;
    private (uint Sequence, int Size)? offer;
    private MemoryStream? incoming;
    private bool active;
    private double requestedAt;
    private int retries;

    public static byte[]? Text(string? value)
    {
        if (value is null || value.Length > MaximumBytes * 2 || value.Contains('\0')) return null;
        try
        {
            var bytes = Utf8.GetBytes(value.Replace("\r\n", "\n").Replace('\r', '\n'));
            return bytes.Length <= MaximumBytes ? bytes : null;
        }
        catch (EncoderFallbackException) { return null; }
    }
    public static string? Decode(byte[] bytes)
    {
        if (bytes.Length > MaximumBytes) return null;
        try { var text = Utf8.GetString(bytes); return text.Contains('\0') || text.Contains('\r') ? null : text; }
        catch (DecoderFallbackException) { return null; }
    }
    public void Reset(uint nextEpoch)
    {
        epoch = nextEpoch; seenOffer = null; local = null; offer = null; incoming?.Dispose(); incoming = null;
        sequence++; retries = 0;
    }
    public void Observe(byte[]? text, bool announce = true)
    {
        local = text is not null && Decode(text) is not null ? text.ToArray() : null;
        sequence++; offer = null; incoming?.Dispose(); incoming = null;
        if (announce) Yield();
    }
    public void Yield() => send(Protocol.Message.ClipGrab, Header(epoch, sequence, local is null ? uint.MaxValue : (uint)local.Length));
    public void SetActive(bool value, double now)
    {
        active = value;
        if (active && offer is not null && incoming is null) Request(now);
    }
    public void Tick(double now)
    {
        if (incoming is null || now - requestedAt < 5) return;
        if (++retries > 3) { offer = null; incoming.Dispose(); incoming = null; return; }
        Request(now);
    }
    public void Receive(Protocol.Message type, byte[] payload, double now)
    {
        if (payload.Length < 12 || (type != Protocol.Message.ClipData && payload.Length != 12))
            throw new InvalidDataException("Invalid clipboard packet length");
        var scope = Read(payload, 0); var id = Read(payload, 4); var value = Read(payload, 8);
        if (scope != epoch) return;
        switch (type)
        {
            case Protocol.Message.ClipGrab:
                if (value != uint.MaxValue && value > MaximumBytes) throw new InvalidDataException("Clipboard too large");
                if (seenOffer == id) return;
                seenOffer = id;
                incoming?.Dispose(); incoming = null;
                offer = value == uint.MaxValue ? null : (id, (int)value);
                retries = 0;
                if (active && offer is not null) Request(now);
                break;
            case Protocol.Message.ClipGet:
                if (id != sequence || local is null) { Yield(); return; }
                if (value > local.Length || value % BlockBytes != 0) throw new InvalidDataException("Invalid clipboard offset");
                send(Protocol.Message.ClipData, Header(epoch, id, value)
                    .Concat(local.Skip((int)value).Take(BlockBytes)).ToArray());
                break;
            case Protocol.Message.ClipData:
                if (payload.Length > BlockBytes + 12) throw new InvalidDataException("Clipboard block too large");
                if (offer is not { } current || current.Sequence != id || incoming is null || value != incoming.Length) return;
                var count = payload.Length - 12;
                if (count != Math.Min(BlockBytes, current.Size - incoming.Length)) throw new InvalidDataException("Invalid clipboard block");
                incoming.Write(payload, 12, count); retries = 0;
                if (incoming.Length < current.Size) { Request(now); return; }
                var bytes = incoming.ToArray();
                incoming.Dispose(); incoming = null; offer = null;
                if (Decode(bytes) is null) throw new InvalidDataException("Invalid clipboard UTF-8");
                local = null; sequence++; // Imported text is never offered back.
                apply(bytes);
                break;
            default: throw new InvalidDataException("Unknown clipboard packet");
        }
    }
    private void Request(double now)
    {
        if (offer is not { } current) return;
        incoming ??= new MemoryStream(current.Size);
        requestedAt = now;
        send(Protocol.Message.ClipGet, Header(epoch, current.Sequence, (uint)incoming.Length));
    }
    public static uint Read(byte[] bytes, int offset) => BinaryPrimitives.ReadUInt32LittleEndian(bytes.AsSpan(offset));
    public static byte[] Header(uint scope, uint id, uint value)
    {
        var bytes = new byte[12];
        BinaryPrimitives.WriteUInt32LittleEndian(bytes, scope);
        BinaryPrimitives.WriteUInt32LittleEndian(bytes.AsSpan(4), id);
        BinaryPrimitives.WriteUInt32LittleEndian(bytes.AsSpan(8), value);
        return bytes;
    }
}

File: DeusKVM/Clipboard/ClipboardTransfer.swift
import Foundation

/// Clipboard blocks are requested individually, leaving room for input/control traffic.
final class ClipboardTransfer {
    static let maximumBytes = 65536
    static let blockBytes = 1024
    var send: (CompanionProtocol.Message, Data) -> Void = { _, _ in }
    var apply: (Data) -> Void = { _ in }
    var onDeadlineChanged: () -> Void = {}
    var retryDeadline: TimeInterval? {
        incoming == nil ? nil : requestedAt + 5
    }

    private var epoch: UInt32 = 0
    private(set) var sequence: UInt32 = 0
    private var local: Data?
    private var seenOffer: UInt32?
    private var offer: (sequence: UInt32, size: Int)?
    private var incoming: Data?
    private var active = false
    private var requestedAt: TimeInterval = 0
    private var retries = 0

    static func text(_ value: String?) -> Data? {
        guard let value, value.utf8.count <= maximumBytes * 2, !value.contains("\0") else { return nil }
        let data = Data(value.replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\r", with: "\n").utf8)
        return data.count <= maximumBytes ? data : nil
    }

    static func decode(_ data: Data) -> String? {
        guard data.count <= maximumBytes, let text = String(data: data, encoding: .utf8),
              !text.contains("\0"), !text.contains("\r") else { return nil }
        return text
    }

    func reset(_ nextEpoch: UInt32) {
        defer { onDeadlineChanged() }
        epoch = nextEpoch; seenOffer = nil; local = nil; offer = nil; incoming = nil; sequence &+= 1; retries = 0
    }

    func observe(_ text: Data?, announce: Bool = true) {
        defer { onDeadlineChanged() }
        local = text.flatMap { Self.decode($0) == nil ? nil : $0 }
        sequence &+= 1; offer = nil; incoming = nil
        if announce { yield() }
    }

    func yield() {
        send(.clipGrab, Self.header(epoch, sequence, local.map { UInt32($0.count) } ?? UInt32.max))
    }

    func setActive(_ value: Bool, now: TimeInterval) {
        defer { onDeadlineChanged() }
        active = value
        if active, offer != nil, incoming == nil { request(now) }
    }

    func tick(_ now: TimeInterval) {
        defer { onDeadlineChanged() }
        guard incoming != nil, now - requestedAt >= 5 else { return }
        retries += 1
        guard retries <= 3 else { offer = nil; incoming = nil; return }
        request(now)
    }

    func receive(_ type: CompanionProtocol.Message, payload: Data, now: TimeInterval) throws {
        defer { onDeadlineChanged() }
        guard payload.count >= 12, type == .clipData || payload.count == 12 else { throw CompanionProtocol.Failure.malformed }
        let scope = CompanionProtocol.read32(payload, 0)
        let id = CompanionProtocol.read32(payload, 4)
        let value = CompanionProtocol.read32(payload, 8)
        guard scope == epoch else { return }
        switch type {
        case .clipGrab:
            guard value == UInt32.max || value <= Self.maximumBytes else { throw CompanionProtocol.Failure.malformed }
            if seenOffer == id { return }
            seenOffer = id
            incoming = nil; retries = 0
            offer = value == UInt32.max ? nil : (id, Int(value))
            if active, offer != nil { request(now) }
        case .clipGet:
            guard id == sequence, let local else { yield(); return }
            guard value <= local.count, value % UInt32(Self.blockBytes) == 0 else { throw CompanionProtocol.Failure.malformed }
            send(.clipData, Self.header(epoch, id, value) + local.dropFirst(Int(value)).prefix(Self.blockBytes))
        case .clipData:
            try receiveBlock(id: id, offset: value, payload: payload, now: now)
        default: throw CompanionProtocol.Failure.malformed
        }
    }

    private func receiveBlock(id: UInt32, offset: UInt32, payload: Data, now: TimeInterval) throws {
        guard payload.count <= Self.blockBytes + 12 else { throw CompanionProtocol.Failure.malformed }
        guard let offer, offer.sequence == id, var bytes = incoming, offset == bytes.count else { return }
        guard payload.count - 12 == min(Self.blockBytes, offer.size - bytes.count) else { throw CompanionProtocol.Failure.malformed }
        bytes.append(payload.dropFirst(12)); incoming = bytes; retries = 0
        if bytes.count < offer.size { request(now); return }
        incoming = nil; self.offer = nil
        guard Self.decode(bytes) != nil else { throw CompanionProtocol.Failure.malformed }
        local = nil; sequence &+= 1
        apply(bytes)
    }

    private func request(_ now: TimeInterval) {
        guard let offer else { return }
        if incoming == nil { incoming = Data() }
        requestedAt = now
        send(.clipGet, Self.header(epoch, offer.sequence, UInt32(incoming?.count ?? 0)))
    }

    static func header(_ epoch: UInt32, _ sequence: UInt32, _ value: UInt32) -> Data {
        CompanionProtocol.u32(epoch) + CompanionProtocol.u32(sequence) + CompanionProtocol.u32(value)
    }
}
